src/Operation/BulkWrite.php

```diff
@@ -191,10 +191,19 @@ public function __construct(string $databaseName, string $collectionName, array
                         throw InvalidArgumentException::invalidType(sprintf('$operations[%d]["%s"][1]', $i, $type), $args[1], 'array or object');
                     }
 
+                    // Treat empty arrays as replacement documents for BC
+                    if ($args[1] === []) {
+                        $args[1] = (object) $args[1];
```

Comment:
I don't get why you need to convert the empty array into an empty object. Replacement documents can be array, isn't it?

Reply:
Without this, `is_pipeline()` with `allowEmpty: true` is going to consider the argument a pipeline and throw an exception (since pipelines are only acceptable for update operations).

Since `is_pipeline()` historically returned `false` for empty arrays, an empty array was always accepted for ReplaceOne and would be encoded as a BSON object in the `update` command (per libmongoc's internal logic discussed in [CDRIVER-4658](https://jira.mongodb.org/browse/CDRIVER-4658)).

Since this PR is adding logic to reject empty pipelines for _both_ update and replace operations, we need to explicitly allow empty arrays for BC.

BulkWrite `replaceOne` operations historically has never used `is_pipeline()`, so pipelines could have snuck through there. Empty PHP arrays would still be encoded as a replacement document, though (same libmongoc logic).

FindOneAndReplace also never used `is_pipeline()` during constructor validation. The only relevant code path was `FindAndModify::createCommandDocument()` conditionally casting the `update` option to an object if `is_pipeline()` returned false (see: [this change](https://github.com/mongodb/mongo-php-library/commit/2628c098e3bfc015d7e2d59d1437cf1fed046e12#diff-efc373289c930c390bc98d4dbbfb00041c37eda8f5689514b6941697f916d770R265-R269) from [PHPLIB-418](https://jira.mongodb.org/browse/PHPLIB-418)). I actually think there was a bug there, as an empty array would have been encoded as a BSON array in the `findAndModify` command -- and thus treated as an empty pipeline (i.e. NOP) in MongoDB 4.2+.

By prohibiting empty pipelines for _both_ updates and replacements, we resolve the last issue with FindOneAndReplace and also enable users to express empty replacement documents as `[]` like they already could for ReplaceOne and BulkWrite `replaceOne` (where they previously might have had to pass `(object) []` explicitly).

Reply:
> I actually think there was a bug there, as an empty array would have been encoded as a BSON array in the findAndModify command -- and thus treated as an empty pipeline (i.e. NOP) in MongoDB 4.2+.

Correction. Since FindOneAndReplace previously never validated its `$replacement` parameter with `is_pipeline()`, an empty BSON array would still have correctly encoded as a BSON document due to this logic: https://github.com/mongodb/mongo-php-library/blob/3a681a3b2f2c0ebac227a3b86bb9057d0e6eb8f8/src/Operation/FindAndModify.php#L295-L299

`is_pipeline()` returns false and the parameter gets cast to an object.

So there was no bug.

Also, I can confirm that the cast in `FindAndModify::createCommandDocument()` is still necessary. If we execute FindAndModify directly (as is done in the test suite), then we rely on that code to ensure an empty array encodes properly.

Technically, we could also change the constructor logic in FindOneAndReplace to cast all `$replacement` arrays to objects `is_pipeline($replacement)` is false, but it's that just adds additional overhead and wouldn't cover the case where FindAndModify is executed directly. The empty array check is cheap and plays nicely with the subsequent validation that rejects any pipeline (empty and non-empty alike).